[MP Rewrite Part 10] Overall improvements to planners

kylestach opened the pull request:
## Description
Improve planners, especially pivot planner. Make angle overrides work correctly.

kylestach committed:
Add angle overrides and fix pivot planner. Small tweaks to settle.

JNeiger commented on soccer/planning/planner/PivotPathPlanner.cpp:
```diff
@@ -71,24 +75,33 @@ Trajectory PivotPathPlanner::plan(const PlanRequest& request) {
     }
 
-    BezierPath pathBezier(points, start_instant.linear_velocity(), Point(0, 0),
+    BezierPath pathBezier(points, Point(0, 0), Point(0, 0),
```
This feels a little weird. What's the reasoning behind this?

kylestach replied:
For pivot, you actually want to explicitly create a circular(ish) path rather than respecting initial velocity.

Without this, when you're constantly replanning the robot just sort of pushes the ball forwards forever.

JNeiger replied:
That makes sense. How well does it work with a fast moving robot coming into a pivot or a robot far away from the pivot?

kylestach replied:
It's a little jank when the robot comes in fast, but it actually gives decent results. It definitely needs to be fixed eventually, if we want to tune motion control really aggressively (which we do). No idea how it works if the robot is very far away from the pivot - with the transition to the action/behavior controller (hey, that's ABC!) we should honestly just fail if we attempt to do a pivot in that context.

JNeiger replied:
Sounds good! So the idea would be to move close to a ball/have it captured, before calling this?

JNeiger commented on soccer/planning/planner/PivotPathPlanner.cpp:
```diff
@@ -71,24 +75,33 @@ Trajectory PivotPathPlanner::plan(const PlanRequest& request) {
+        RobotInstant instant = path.last();
+        instant.stamp = instant.stamp + RJ::Seconds(5.0);
+        path.AppendInstant(instant);
+    }
```
This might be a visual glitch because I'm on mobile. These braces seem useless?

kylestach replied:
They're useless, I like to do this to make a temporary that shouldn't be used anywhere else. It's just personal preference, I can get rid of it if you'd rather not have it.

JNeiger replied:
I gotcha. Personally I'm not a fan of it because it messes with the tabbing whitespace, but I'm also not doing much coding anymore sooooo it's up to you guys

oswinso replied:
I think its a bit of visual clutter. I'm not a big fan of it either

kylestach replied:
Fixed in #1542 by deleting this code anyway, lmao

JNeiger commented on soccer/planning/planner/SettlePlanner.cpp:
```diff
@@ -282,10 +281,18 @@ Trajectory SettlePlanner::intercept(
         }
     }
 
+    Geometry2d::Point ballVelIntercept;
```
Can you throw in a comment about the edge case this will be triggered? I assume it's to deal with the case the ball stops before we would reach it

kylestach commented on soccer/Processor.cpp:
```diff
@@ -268,14 +266,16 @@ void Processor::run() {
-        // Log this entire frame
-        _logger->run();
+        {
+            loopMutex()->lock();
```
This just keeps the UI from locking up quite as aggressively. It should probably be reverted.

kylestach committed:
Merge branch 'mp-rewrite/automated-tests' into mp-rewrite/final-planner-improvements

kylestach committed:
Clean up and comment logic in settle planner

JNeiger reviewed (approved):
Nothing blocking, LGTM from my pov

oswinso commented on soccer/planning/planner/PivotPathPlanner.cpp:
```diff
@@ -39,12 +39,15 @@ Trajectory PivotPathPlanner::plan(const PlanRequest& request) {
+    double radius = _pivotRadiusMultiplier->value() * Robot_Radius;
+    auto pivot_point = command.pivotPoint;
+
+    if (cached_pivot_point.has_value() && cached_pivot_point.value().distTo(pivot_point) < Robot_MouthWidth / 2) {
```
What's the logic here? Why mouthwidth / 2?

JNeiger replied:
The usual thought is that if you did the calculations perfectly, you have about this much wiggle room before you won't hit the ball with the mouth. Replanning for micrometers isn't worth it since replanning caused more performance problems than than being a little off historically.

This measurement falls apart a little when you have to worry about forwards backwards motion, but the thought process was always kinda "meh, there's worse things causing problems"

oswinso commented on soccer/planning/planner/PivotPathPlanner.hpp:
```diff
@@ -17,11 +17,16 @@ class PivotPathPlanner : public PlannerForCommandType<PivotCommand> {
-    [[nodiscard]] bool shouldReplan(const PivotCommand& command) const;
     Trajectory previous;
+
+    std::optional<Geometry2d::Point> cached_pivot_point;
```
A comment about when this is cached / invalidated, and why we need to cache maybe?

oswinso reviewed (approved):
lgtm

kylestach committed:
Merge branch 'mp-rewrite/automated-tests' into mp-rewrite/final-planner-improvements

kylestach committed:
Make settle planner pick best buffer

This fixes the edge case where the planner can't find an path that finds
the ball quickly enough because the only way to catch it doesn't have
enough buffer time (i.e. immediately before it is caught).

kylestach committed:
[MP Rewrite Part 11] Miscellaneous formatting and cleanup (#1540)

* Formatting

* More formatting

* Tidy fixes

* Straggling formatting

* Formatting

* Rename low_level to primitives

* Comment on cached_pivot_point

[MP Rewrite Part 12] Port the line kick planner (#1541)

* Add in sketchy line kick

* Formatting

* Tidy

[MP Rewrite Part 13] Stop play fixes and holding trajectory (#1542)

* Fix stopped play and hold in place at end of path move

* Formatting

* Dumb NOLINT

* Style again
